File: labs/13-a2a-agent/python/main.py
"""Minimal A2A-style Agent Card, Task submission, and status polling.

Runs a small HTTP agent server and a client that discovers it, submits a task,
and polls the task status until completion.
"""

import json
import logging
import os
import threading
import time
import uuid
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
logger = logging.getLogger("a2a-agent")

AGENT_CARD = {
    "name": "agent-labs-echo-agent",
    "description": "A minimal A2A agent that echoes input after a short delay.",
    "url": "http://localhost:8123",
    "version": "0.1.0",
    "capabilities": {
        "streaming": False,
        "pushNotifications": False,
    },
    "skills": [
        {
            "id": "echo",
            "name": "Echo",
            "description": "Returns the input text as the task result.",
        }
    ],
}

TASKS: dict[str, dict[str, Any]] = {}
TASK_LOCK = threading.Lock()


def create_task(message: dict[str, Any]) -> dict[str, Any]:
    task_id = str(uuid.uuid4())
    task = {
        "id": task_id,
        "status": {"state": "submitted", "timestamp": time.time()},
        "messages": [message],
        "artifacts": [],
    }
    with TASK_LOCK:
        TASKS[task_id] = task

    # Simulate async work in a background thread.
    def worker():
        time.sleep(2)
        with TASK_LOCK:
            task["status"] = {"state": "working", "timestamp": time.time()}
        time.sleep(2)
        with TASK_LOCK:
            task["status"] = {
                "state": "completed",
                "timestamp": time.time(),
                "message": {
                    "role": "agent",
                    "parts": [{"type": "text", "text": f"Echo: {message.get('parts', [{}])[0].get('text', '')}"}],
                },
            }

    threading.Thread(target=worker, daemon=True).start()
    return task


class AgentHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        logger.info(format % args)

    def _send_json(self, status: int, payload: Any) -> None:
        body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == "/.well-known/agent.json":
            self._send_json(200, AGENT_CARD)
            return

        if parsed.path.startswith("/tasks/"):
            task_id = parsed.path.split("/")[-1]
            with TASK_LOCK:
                task = TASKS.get(task_id)
            if task:
                self._send_json(200, task)
            else:
                self._send_json(404, {"error": "Task not found"})
            return

        self._send_json(404, {"error": "Not found"})

    def do_POST(self):
        parsed = urlparse(self.path)
        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length).decode("utf-8")
        try:
            payload = json.loads(body) if body else {}
        except json.JSONDecodeError:
            self._send_json(400, {"error": "Invalid JSON"})
            return

        if parsed.path == "/tasks/send":
            message = payload.get("message", {})
            task = create_task(message)
            self._send_json(200, task)
            return

        self._send_json(404, {"error": "Not found"})


def start_server(host: str = "localhost", port: int = 8123) -> HTTPServer:
    server = HTTPServer((host, port), AgentHandler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    logger.info("Agent server listening on http://%s:%d", host, port)
    return server


def fetch_agent_card(base_url: str) -> dict[str, Any]:
    import urllib.request

    with urllib.request.urlopen(f"{base_url}/.well-known/agent.json") as resp:
        return json.loads(resp.read().decode("utf-8"))


def submit_task(base_url: str, text: str) -> dict[str, Any]:
    import urllib.request

    payload = {
        "message": {
            "role": "user",
            "parts": [{"type": "text", "text": text}],
        }
    }
    req = urllib.request.Request(
        f"{base_url}/tasks/send",
        data=json.dumps(payload).encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode("utf-8"))


def get_task(base_url: str, task_id: str) -> dict[str, Any]:
    import urllib.request

    with urllib.request.urlopen(f"{base_url}/tasks/{task_id}") as resp:
        return json.loads(resp.read().decode("utf-8"))


def poll_task(base_url: str, task_id: str, timeout: float = 30.0, interval: float = 0.5) -> dict[str, Any]:
    deadline = time.time() + timeout
    while time.time() < deadline:
        task = get_task(base_url, task_id)
        state = task.get("status", {}).get("state")
        logger.info("Task %s state: %s", task_id[:8], state)
        if state in ("completed", "failed"):
            return task
        time.sleep(interval)
    raise TimeoutError(f"Task {task_id} did not complete within {timeout}s")


def main():
    base_url = os.getenv("A2A_AGENT_URL", "http://localhost:8123").rstrip("/")
    own_server = None

    try:
        # If the configured agent is not reachable, start one locally.
        try:
            fetch_agent_card(base_url)
            logger.info("Using existing agent at %s", base_url)
        except Exception:
            logger.info("No agent found at %s; starting local agent", base_url)
            own_server = start_server()
            # Wait briefly for the server to be ready.
            for _ in range(20):
                try:
                    fetch_agent_card(base_url)
                    break
                except Exception:
                    time.sleep(0.1)

        card = fetch_agent_card(base_url)
        print("\n[Agent Card]")
        print(json.dumps(card, indent=2, ensure_ascii=False))

        task = submit_task(base_url, "Hello from the A2A client!")
        print("\n[Submitted Task]")
        print(json.dumps(task, indent=2, ensure_ascii=False))

        final = poll_task(base_url, task["id"])
        print("\n[Final Task]")
        print(json.dumps(final, indent=2, ensure_ascii=False))

    finally:
        if own_server:
            own_server.shutdown()


if __name__ == "__main__":
    main()

File: labs/13-a2a-agent/node/main.ts
import http from "http";
import { randomUUID } from "crypto";

const AGENT_CARD = {
  name: "agent-labs-echo-agent",
  description: "A minimal A2A agent that echoes input after a short delay.",
  url: "http://localhost:8123",
  version: "0.1.0",
  capabilities: {
    streaming: false,
    pushNotifications: false,
  },
  skills: [
    {
      id: "echo",
      name: "Echo",
      description: "Returns the input text as the task result.",
    },
  ],
};

const TASKS: Record<string, any> = {};
const TASK_LOCK: Promise<void> = Promise.resolve();

async function withLock<T>(fn: () => T): Promise<T> {
  return await TASK_LOCK.then(fn);
}

function createTask(message: any): any {
  const taskId = randomUUID();
  const task: any = {
    id: taskId,
    status: { state: "submitted", timestamp: Date.now() },
    messages: [message],
    artifacts: [],
  };
  TASKS[taskId] = task;

  setTimeout(() => {
    task.status = { state: "working", timestamp: Date.now() };
  }, 2000);
  setTimeout(() => {
    const text = message.parts?.[0]?.text || "";
    task.status = {
      state: "completed",
      timestamp: Date.now(),
      message: {
        role: "agent",
        parts: [{ type: "text", text: `Echo: ${text}` }],
      },
    };
  }, 4000);

  return task;
}

function sendJson(res: http.ServerResponse, status: number, payload: any): void {
  const body = JSON.stringify(payload);
  res.writeHead(status, { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(body) });
  res.end(body);
}

function startServer(host = "localhost", port = 8123): http.Server {
  const server = http.createServer((req, res) => {
    const url = new URL(req.url || "/", `http://${req.headers.host}`);
    if (req.method === "GET" && url.pathname === "/.well-known/agent.json") {
      sendJson(res, 200, AGENT_CARD);
      return;
    }
    if (req.method === "GET" && url.pathname.startsWith("/tasks/")) {
      const taskId = url.pathname.split("/").pop();
      const task = taskId ? TASKS[taskId] : null;
      if (task) sendJson(res, 200, task);
      else sendJson(res, 404, { error: "Task not found" });
      return;
    }
    if (req.method === "POST" && url.pathname === "/tasks/send") {
      let body = "";
      req.on("data", (chunk) => (body += chunk));
      req.on("end", () => {
        try {
          const payload = body ? JSON.parse(body) : {};
          const task = createTask(payload.message || {});
          sendJson(res, 200, task);
        } catch {
          sendJson(res, 400, { error: "Invalid JSON" });
        }
      });
      return;
    }
    sendJson(res, 404, { error: "Not found" });
  });
  server.listen(port, host, () => {
    console.log(`Agent server listening on http://${host}:${port}`);
  });
  return server;
}

function fetchJson(url: string, options: http.RequestOptions & { body?: string } = {}): Promise<any> {
  return new Promise((resolve, reject) => {
    const req = http.request(url, options, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try {
          resolve(JSON.parse(data));
        } catch {
          resolve(data);
        }
      });
    });
    req.on("error", reject);
    if (options.body) req.write(options.body);
    req.end();
  });
}

async function fetchAgentCard(baseUrl: string): Promise<any> {
  return fetchJson(`${baseUrl}/.well-known/agent.json`);
}

async function submitTask(baseUrl: string, text: string): Promise<any> {
  const payload = {
    message: {
      role: "user",
      parts: [{ type: "text", text }],
    },
  };
  return fetchJson(`${baseUrl}/tasks/send`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

async function getTask(baseUrl: string, taskId: string): Promise<any> {
  return fetchJson(`${baseUrl}/tasks/${taskId}`);
}

async function pollTask(
  baseUrl: string,
  taskId: string,
  timeout = 30000,
  interval = 500
): Promise<any> {
  const deadline = Date.now() + timeout;
  while (Date.now() < deadline) {
    const task = await getTask(baseUrl, taskId);
    const state = task.status?.state;
    console.log(`Task ${taskId.slice(0, 8)} state: ${state}`);
    if (state === "completed" || state === "failed") return task;
    await new Promise((r) => setTimeout(r, interval));
  }
  throw new Error(`Task ${taskId} did not complete within ${timeout}ms`);
}

async function main() {
  const baseUrl = (process.env.A2A_AGENT_URL || "http://localhost:8123").replace(/\/$/, "");
  let ownServer: http.Server | null = null;

  try {
    try {
      await fetchAgentCard(baseUrl);
      console.log(`Using existing agent at ${baseUrl}`);
    } catch {
      console.log(`No agent found at ${baseUrl}; starting local agent`);
      ownServer = startServer();
      for (let i = 0; i < 20; i++) {
        try {
          await fetchAgentCard(baseUrl);
          break;
        } catch {
          await new Promise((r) => setTimeout(r, 100));
        }
      }
    }

    const card = await fetchAgentCard(baseUrl);
    console.log("\n[Agent Card]");
    console.log(JSON.stringify(card, null, 2));

    const task = await submitTask(baseUrl, "Hello from the A2A client!");
    console.log("\n[Submitted Task]");
    console.log(JSON.stringify(task, null, 2));

    const final = await pollTask(baseUrl, task.id);
    console.log("\n[Final Task]");
    console.log(JSON.stringify(final, null, 2));
  } finally {
    if (ownServer) ownServer.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
